crates/metal-llvm/src/codegen.rs

```diff
@@ -32,66 +19,92 @@ pub struct CodeGen {
 }
 
 impl CodeGen {
-    // complete codegen process
-    pub fn finish(&self) -> &'static str {
-        unsafe {
-            let ir = LLVMPrintModuleToString(self.module);
-            let cir = CStr::from_ptr(ir);
-            // TODO: specify which module specifically.
-            let string_ir = cir
-                .to_str()
-                .expect("Failed to turn LLVM module into string");
-            string_ir
+    pub fn get_types(&self, types: Vec<Type>) -> Vec<LLVMTypeRef> {
+        let mut v = Vec::with_capacity(types.len());
+        for t in types {
+            v.push(self.get_type(t))
         }
+        v
     }
 
-    pub fn primitive(&self, primitive: &metal_ast::Primitives) -> LLVMTypeRef {
-        match primitive {
-            metal_ast::Primitives::I8 => unsafe { LLVMIntTypeInContext(self.ctx, 8) },
-            metal_ast::Primitives::I16 => unsafe { LLVMIntTypeInContext(self.ctx, 16) },
-            metal_ast::Primitives::I32 => unsafe { LLVMIntTypeInContext(self.ctx, 32) },
-            metal_ast::Primitives::I64 => unsafe { LLVMIntTypeInContext(self.ctx, 64) },
-            metal_ast::Primitives::I128 => unsafe { LLVMIntTypeInContext(self.ctx, 128) },
-
-            metal_ast::Primitives::F32 => unsafe { LLVMFloatTypeInContext(self.ctx) },
-            metal_ast::Primitives::F64 => unsafe { LLVMDoubleTypeInContext(self.ctx) },
-            metal_ast::Primitives::F128 => unsafe { LLVMFP128TypeInContext(self.ctx) },
-
-            metal_ast::Primitives::String => {
-                todo!()
-            }
-            metal_ast::Primitives::Void => unsafe { LLVMVoidTypeInContext(self.ctx) },
+    pub fn get_type(&self, t: metal_mir::types::Type) -> LLVMTypeRef {
+        match t {
+            metal_mir::types::Type::Primitive(p) => unsafe {
+                self.get_llvm_type_primitive(*(p.clone()))
+            },
+            metal_mir::types::Type::Composite(c) => unsafe {
+                match *c {
+                    metal_mir::types::Composite::Tuple(t) => LLVMStructTypeInContext(
+                        self.ctx,
+                        self.get_types(t.types.clone()).as_mut_ptr(),
```

Comment:
Why the clone?

Reply:
Then the length call should be extracted into a variable instead of cloning the vec